### PR TITLE
Fixed Point Logic and Fixed Point to Floating Point Conversion

### PR DESCRIPTION
**Pull request by soneryaldiz**
## Description & Motivation

This PR introduces a new signal type `FixedPoint` for fixed point arithmetic, and adds a generic module that can convert a fixed-point input to floating-point output with rounding to nearest even.

## Testing

Testing includes:
- Exhaustive conversion from signed Q4.4 to E3M2.
- Exhaustive conversion from unsigned Q4.4 to E3M2.
- Exhaustive conversion from signed Q0.8 to E3M2 (shrinking precision).
- Exhaustive conversion from signed Q0.3 to E5M6 (expanding precision).

## Backwards-compatibility

> Is this a breaking change that will not be backwards-compatible? If yes, how so?

Not a braking change.

## Documentation

> Does the change require any updates to documentation? If so, where? Are they included?

Documentation added to `fixed_point.md`.

### CODE REVIEW

File: lib/src/arithmetic/fixed_to_float.dart [hunk 1]
```diff
@@ -0,0 +1,188 @@
+// Copyright (C) 2023-2024 Intel Corporation
```
Comment (mkorbel1) [inline lib/src/arithmetic/fixed_to_float.dart:1]: probably just 2024? :)

File: lib/src/arithmetic/signals/fixed_point_logic.dart [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        throw RohdHclException('Value is not compatible with signal.');
+      }
+      super.put(val.value);
+    } else {
+      throw RohdHclException('Only FixedPointValue is allowed');
+    }
+  }
+
+  /// Check compatibility
+  void _compatible(dynamic other) {
```
Comment (mkorbel1) [inline lib/src/arithmetic/signals/fixed_point_logic.dart:68]: maybe call it something like `_verifyCompatible`?

**Comment by soneryaldiz** [comment]
The component supports both normal and subnormal now. I will continue with documentation and configuration.

File: test/arithmetic/fixed_to_float_test.dart [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+// 2024 October 24
+// Author: Soner Yaldiz <soner.yaldiz@intel.com>
+
+import 'dart:io';
+import 'dart:math';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+import 'package:test/test.dart';
+
+void main() async {
```
Comment (mkorbel1) [inline test/arithmetic/fixed_to_float_test.dart:16]: maybe some more tests to cover corner cases? sweep exhaustive for some small low precision if we can?  normals/subnormals etc?

Comment (soneryaldiz) [reply test/arithmetic/fixed_to_float_test.dart:16]: Please see `'Q4.4 to E3M2' below. Exhaustive test from '0 to '1.

Comment (mkorbel1) [reply test/arithmetic/fixed_to_float_test.dart:16]: oh you're right, my bad, nice!

File: lib/src/arithmetic/fixed_to_float.dart [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    addOutput('float', width: _float.width) <= _float;
+
+    final bias = FloatingPointValue.computeBias(exponentWidth);
+    final eMax = pow(2, exponentWidth) - 2;
+    final iWidth =
+        (2 + max(fixed.n, max(log2Ceil(fixed.width), exponentWidth))).toInt();
+
+    // Special handling needed for E4M3 as it does not support inf
+    if ((exponentWidth == 4) && (mantissaWidth == 3)) {
+      UnimplementedError('E4M3 is not supported.');
```
Comment (mkorbel1) [inline lib/src/arithmetic/fixed_to_float.dart:46]: It's probably a good idea to explicitly note this in the constructor doc string rather than let people discover it after using it

File: lib/src/component_config/components/config_fixed_to_float.dart
Comment (mkorbel1) [inline lib/src/component_config/components/config_fixed_to_float.dart:1]: add this component to the `component_registry.dart` as well

File: test/arithmetic/fixed_to_float_test.dart [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+  test('Smoke', () async {
+    final fixed = FixedPoint(signed: true, m: 34, n: 33);
+    final dut = FixedToFloat(fixed, exponentWidth: 8, mantissaWidth: 3);
+    await dut.build();
+    File('${dut.name}.sv').writeAsStringSync(dut.generateSynth());
```
Comment (mkorbel1) [inline test/arithmetic/fixed_to_float_test.dart:21]: do not leave these file-write operations in the test -- after test run there should be no temporary files remaining

File: test/arithmetic/fixed_to_float_test.dart [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+          exponentWidth: dut.exponentWidth, mantissaWidth: dut.mantissaWidth);
+      expect(fpv.sign, fpvExpected.sign);
+      expect(fpv.exponent.bitString, fpvExpected.exponent.bitString,
+          reason: 'exponent mismatch');
+      expect(fpv.mantissa.bitString, fpvExpected.mantissa.bitString,
+          reason: 'mantissa mismatch');
+    }
+  });
+
+  // Test is skipped as FloatingPointValue.ofDouble does not handle infinities.
```
Comment (mkorbel1) [inline test/arithmetic/fixed_to_float_test.dart:144]: you could mark this as a `// TODO(owner): ... <link to bug>` so that we know to enable the test once the issue is fixed.  it's also worth adding a reminder in the github issue so whenever it is fixed we remember to enable this

**Review by mkorbel1 — CHANGES_REQUESTED** [review]
Looks great!!  Few small things then ready to merge I think

File: test/arithmetic/fixed_to_float_test.dart (hunk 5 above)
Comment (soneryaldiz) [reply test/arithmetic/fixed_to_float_test.dart:21]: Should `dut.build()` be kept? Does it catch any issue that is missed by a test?

**Comment by soneryaldiz** [comment]
Resolved comments.

File: test/arithmetic/fixed_to_float_test.dart (hunk 5 above)
Comment (mkorbel1) [reply test/arithmetic/fixed_to_float_test.dart]: It's fine to keep, it may help catch some issues like input port rule violations

**Review by mkorbel1 — APPROVED** [review]
Great job!
